src/main/BlockPool.cpp

```diff
@@ -0,0 +1,1216 @@
+#define MAX_QUARANTINE_SIZE (10<<20)
+
+// Redzones are added before and after the allocation.
+#define REDZONE_SIZE (16)
+
+// Quarantines are used so objects are not quickly reused after free.
+static size_t small_quarantine_size = 0;
+static size_t quarantine_size = 0;
+static FreeListNode* small_quarantine[NUM_SMALL_POOLS];
+static FreeListNode* quarantine[64];
+
+// Helper function for address calculations.
+static void* offset_pointer(void* pointer, size_t bytes) {
+    return static_cast<char*>(pointer) + bytes;
+}
+
+#endif // HAVE_ADDRESS_SANITIZER
+
+/**
+ * Hash bucket for the allocation table.
+ * The two lowest bits of the data pointer are flags indicating
+ * if the hash bucket contains a superblock (2) and whether this
+ * hash bucket is the first representing that allocation (1).
+ */
+struct HashBucket {
+  uintptr_t data;
+  unsigned size;
+
+  /* Masks away the pointer bits from the data field. */
+  uintptr_t DataPointer() {
+    return data & ~uintptr_t{3};
+  }
+};
+
+struct Superblock;
+
+/**
+ * Freelists are used to store superblock nodes and large allocations.
+ * For large allocations the block and superblock members are unused.
+ */
+struct FreeListNode {
+  FreeListNode* next;
+  u32 block;  // Used only for superblock free nodes.
+  Superblock* superblock;  // Used only for superblock free nodes.
+};
+
+/**
+ * Superblocks are used to allocate small and medium sized objects.
+ * The bitset tracks which blocks are currently allocated.
+ * For small objects block_size is the actual object size.
+ * For medium objects block_size is the 2-log of the object size.
+ */
+struct Superblock {
+  u32 block_size;
+  u32 next_untouched;
+  FreeListNode* free_list;
+  u64 free[];  // Free bitset.
+
+  Superblock(u32 block_size, unsigned bitset_entries):
+      block_size(block_size),
+      next_untouched(0),
+      free_list(nullptr) {
+    // Here we mark all bitset entries as free, later we don't have to do
+    // precise range checking while iterating allocated blocks when
+    // the number of blocks is not evenly divisible by 64.
+    for (int i = 0; i < bitset_entries; ++i) {
+      free[i] = ~0ull;
+    }
+  }
+
+  /**
+   * Allocates a new superblock from the small object arena.
+   * Returns nullptr if the arena space is full.
+   * If nullptr is returned we need to fall back on using a separately
+   * allocated superblock.
+   */
+  static Superblock* NewSuperblock(int block_size);
+
+  /**
+   * Compute the address of the superblock header from a pointer.
+   * Returns nullptr if candidate pointer is not inside the small object arena
+   * or if the pointer points inside the superblock header.
+   */
+  static Superblock* SuperblockFromPointer(void* pointer);
+
+  /**
+   * Allocate a small block (bytes >= 32 && bytes <= 256).
+   * Returns nullptr if there is no more space left for this object
+   * size in the small object arena.
+   */
+  static void* AllocSmall(size_t block_size);
```

Comment:
This should be a quick function that then dispatches to Superblock::AllocateSmallBlock()